README.md
<div align="center">

![Foto](https://github.com/lmistie/VoiceHelper_Margo/blob/main/src/Margo.gif)
</div>

* [Логика проекта](#Project)
* [Команды для Марго](#Commands) 
* [Встречи](#Meetings) 
* [Code review](#Code_review) 
* [Библиотеки](#Libraries)


<h2 align="center"><font size="8px">Логика проекта <a name="Project"></a></font></h2>
<p align="center">Сама идея пришла к нам не сразу.<br>Изначально главным аспектам проекта было создание, голосового Ассистента, который сможет выделяться на фоне остальных Голосовых Помощников (ГП).<br>Главной проблемой современных ГП является их составляющие ядро, кто-то запускается из браузера, кто-то на определённой Операционной системе, кто-то требует покупку дополнительного оборудования.
Тем самым мы команда начинающих разработчиков, решили рискнуть и сделать Универсального Голосового Помощника Марго.
Огромным преимуществом нашего продукта является простота и удобно проекта.Наш голосовой ассистент	 будет работать: в сфере мобильных телефонов, ПК и ноутбуки любых ОС, Машины и Электрокары, ТВ, Бытовая техника и умный дом и это еще не все идеи которые мы готовы воплотить.</p>

![Foto](https://github.com/lmistie/VoiceHelper_Margo/blob/main/src/scheme.jpeg)

<p align="center"><font size="3px"> Пока весь проект написан с помощью:</font></p>
<div align="center">

![Python](https://img.shields.io/badge/-Python-000000?style=flat-square&logo=python&logoWidth=20)
![Vscode](https://img.shields.io/badge/-VScode-000000?style=flat-square&logo=VisualStudioCode&logoWidth=20&logoColor=blue)
![Vim](https://img.shields.io/badge/-Vim-000000?style=flat-square&logo=vim&logoWidth=20&logoColor=green)
![Notepad++](https://img.shields.io/badge/-Notepad++-000000?style=flat-square&logo=Notepad%2b%2b&logoWidth=20&logoColor=green)
![JetBrains](https://img.shields.io/badge/-JetBrains-000000?style=flat-square&logo=JetBrains&logoWidth=20&logoColor=green)
![iTerm2](https://img.shields.io/badge/-iTerm2-000000?style=flat-square&logo=iTerm2&logoWidth=20)
![PowerShell](https://img.shields.io/badge/-PowerShell-000000?style=flat-square&logo=PowerShell&logoWidth=20)
</div>
<!--<p align="center"><font size="3px">Дальше планируется добавление других языков программирования</font></p>-->

<h5 align="center"><font size="3px">Описание умений и команд голосового помощника Марго<a name="Commands"></a></font></h5>

<div align="center">

| Команды | Работоспасобность |
|:------------:|:------------:|
| Привет | ❌ |
| Пока | ✅ |
| Выключи комп | ✅ |
| Открой Google | ✅ |
| Открой Vkontakte | ✅ |
| Открой Instagram | ✅ |
| Включи YouTube | ✅ |
| Покажи погоду | ✅ |
| Очисти историю| ❌ |
| Покажи файл со статистикой | ❌ |
| Переведи слово | ✅  |
| Открой список команд | ❌ |
| Какие планы | ❌ |
</div>
<br>
<p align="center"><font size="3px">Этапы развития встречи 15.08.21</font></p><a name="Meetings"></a>
<div align="center">

| № | Задачи |
|:---:|:------------:|
| 1️⃣ | Пишем в любом удобном редакторе и под любое ОС |
| 2️⃣ | Придерживаемся станарта PEP 8 по коду в проекте |
| 3️⃣ | Новые фичи добавляем тегами или в новую ветку проекта | 
| 4️⃣ | Первостепенная задача: <br>❗ Правильное распознавание речи голосовым помощником<br>❗ Правильная обработка речи и преобразование речи в запрос |
</div>
<br>
</font></p>

<p align="center">Code Review<a name="Code_review"></a></p>

|  | Задачи на исправления (ММ) от 20.09.2021|
|:---:|:------------:| 
| 1 | Требуется логирование ошибок:<br> При каждом запуске создавать в папке logs файл именуемый текущей датой и временем. <br>В файл писать все exception, ошибки и логи для удобства дебага/разработки.<br>По хорошему создавать два файла, в одном все логи, в другом только `errors/exceptions`.<br>@lmistie🧑🏻‍💻 |
| 2 | Переделать функцию чтения истории файлов, на данный момент функция неправильно работает с массивами/словарями.<br>На мой взгляд сложная конструкция.<br>Возможно историю запроса нужно писать в csv файл, в котором следующий формат<br>`"timestamp" "phrase"`.<br>@lmistie🧑🏻‍💻 |
| 3 | Подтянуть распознавание речи, на данный момент работает с переменным успехом.<br>@den_mko🧑🏻‍💻 |
| 4 | Вынести все возможные фразы в отдельный файл(сейчас в коде), например в формате `json`.<br>@lmistie🧑🏻‍💻 |
| 5 | Попробовать собрать venv.<br>Протестировать либы, на виндах и никсах (возможен ли одинаковый venv для различных платформ).<br>Перед запуском *****.py, необходимо активировать venv: `./venv/scripts/activate`<br>@lmistie🧑🏻‍💻 и @den_mko🧑🏻‍💻 |
| 6 | Для ускорения процесса разработки дальнейшего функционала обработки,<br> сделать возможность текстового ввода или ввода из файла,<br> чтобы не привязываться только к голосу с кривым, на данный момент, распознаванием.|
<!--|  | Задачи на исправления (ММ) от 20.09.2021| -->
<br>

<p align="center"><font size="3px">Описание библеотек и их назначения</font></p><a name="Libraries"></a>

| Описание библеотек: | Работа в offline-режиме | Требуемые зависимости |
|:-------:|:----------------:|:---------------:|
| Распознавать и синтезировать речь: | ✅ | [pip install PyAudio](https://pypi.org/project/PyAudio/)<br>[pip install pyttsx3](https://pypi.org/project/pyttsx3/)<br>[pip install SpeechRecognition](https://pypi.org/project/SpeechRecognition/)<br>[pip install vosk](https://pypi.org/project/vosk/)|
| Сообщать о прогнозе погоды в любой точке мира: | ❌ | [pip install pyowm](https://pypi.org/project/pyowm/) (OpenWeatherMap) |
| Производить поисковый запрос в поисковой системе Google (а также открывать сами результаты данного запроса) | ❌ | [pip install google](https://pypi.org/project/google/)|
| Производить поисковый запрос видео в системе YouTube | ❌ | Пока нет решения |
| Выполнять поиск определения в Wikipedia c дальнейшим прочтением первых двух предложений | ❌ | [pip install wikipedia-api](https://pypi.org/project/Wikipedia-API/)|
| Переводить фразы с изучаемого языка на родной язык пользователя и наоборот | ❌ | [pip install googletrans](https://pypi.org/project/googletrans/)<br> (Google Translate)|
| Искать человека по имени и фамилии в социальных сетях | ❌ | Пока нет решения |
| «Подбрасывать монетку» | ✅ | Пока нет решения |
| Здороваться и прощаться (после прощания работа приложения завершается) | ✅ | Пока нет решения |
| Менять настройки языка распознавания и синтеза речи на ходу | ✅ | Пока нет решения |

history.txt
переведи
переведи
переведи
переведи
переведи
до скорого
какие планы
покажи файл статистики
переведи
переведи
открой список команд
какая сейчас погода


pyapp.py
# BETA ВЕРСИЯ ГОЛОСОВОГО ПОМОЩНИКА МАРГО V 1.0

# импортирование библиотек
import speech_recognition as sr
import os
import sys
import webbrowser

def talk(words):
    print(words)
    os.system("say " + words)

talk("Привет, я ваш голосовой помощник. Иногда я торможу, потому что я еще в разработке. Ну что приступим?")

def command():
    r = sr.Recognizer()

    with sr.Microphone() as source:
        print("Говорите")
        r.pause_threshold = 1
        r.adjust_for_ambient_noise(source, duration=1)
        audio = r.listen(source)
    try:
        tz = r.recognize_google(audio, language="ru-RU").lower()
        print("Вы сказали " + tz)
    except sr.UnknownValueError:
        talk("Я вас не поняла")
        tz = command()

    return tz

def makeSomething(tz):
    if 'открой google' in tz:
        talk("Секундочку")
        url_0 = 'https://www.google.com'
        webbrowser.open(url_0)
    elif 'открой яндекс' in tz:
        talk("Секундочку")
        url_1 = 'https://yandex.ru/'
        webbrowser.open(url_1)
    elif 'как тебя зовут' in tz:
        talk("Меня зовут Марго")
    elif 'что ты умеешь делать' in tz:
        talk("Пока я могу назвать свое имя и открывать веб-сайты")
    elif 'марго' in tz:
        talk("Я вас слушаю, говорите")
    elif 'спасибо стоп' in tz:
        talk("Хорошо, до новых встреч, пока")
        sys.exit()

while True:
    makeSomething(command())

vh-1.0.py
# импортирование библиотек
import speech_recognition as sr
import os
import sys
import webbrowser
import random
import pyttsx3
from tkinter import *
import pandas as pd
from fuzzywuzzy import fuzz
import time
from datetime import datetime
from colorama import *

# Глобальные переменные
text = '' # Весь текст
r = sr.Recognizer() # Микрофон на устройсте, который слушает речь
engine = pyttsx3.init(); # Движок
adr = '' # Google браузер
i = 0 # для счетчика
task_number = 0


def speak(words):
    print(words)
    os.system("say " + words)


# Словарь приветствия Марго
name_voice = ['марго', 'маргорита', 'рита', 'ритка', 'маргунка', 'подруга' ,'не могла бы ты', 'пожалуйста', 'сделай услугу', 'покажи ка','покажи ']
# Словарь возможных команд (должно быть много пока без ИИ)
command = ['привет','здорова','выключи компьютер','выключи комп','включи youtube','открой браузер','открой google','открой список команд','какие команды ты знаешь','открой vk','открой instagram','открой stackoverflow','покажи погоду','какая сейчас погода','покажи погоду на завтра','очисти файл','очисти файл статистики','покажи статистику','покажи файл со статистикой','покажи файл статистики','покажи стату','какие планы','пока','до встречи','еще увидимся','до скорого','хорошего дня','хорошего вечера']

# Раздел функций с описанием команд Марго(Почитайте про Append и как работает библиотека pandas [53 строка pd.DataFrame] !!!!! )

# История запросов
def history():
    z = {}
    arr = {}
    arr2 = {}
    arr3 = {}
    arr4 = {}
    file = open('history.txt', 'r', encoding = 'UTF-8')
    case = file.readlines()
    for i in range(len(case)):
        line = str(case[i].replace('\n','').strip())
        arr.dict(line)
    file.close()
    for i in range(len(arr)):
        x = arr[i]
        if x in z:
            z[x] += 1
        if not(x in z):
            b = {x : 1}
            z.update(b)
        if not(x in arr2):
            arr2.append(x)
    for i in arr2:
        arr3.append(z[i])
    for i in range(1, len(arr3)+1):
        arr4.append(str(i)+') ')
    list = pd.DataFrame({
        'command' : arr2,
        'count' : arr3
    }, index = arr4)
    list.index.name = '№'
    print(list)

# Планы на проект Марго 1.0
def plane():
    global engine
    plane = '''Привет, я ваш голосовой помощник. Иногда я торможу, потому что я еще в разработке. В будующем я планирую быть конкурентом и возможно ассистентом вашего умного дома. Начиная от вашей технике, заканчивая вашим автомобилем и бытовой техникой. Пока еще у меня не было нормальных тестов, так что не ругайтесь, я все еще учусь.'''
    engine.say(plane)

# Очистка файла со статистикой
def clear_static():
    global engine
    file = open('history.txt', 'w', encoding= 'UTF-8') # "w" - Запись - Открывает файл для записи, создает файл, если он не существует
    file.close()
    engine.say('Ваш файл со статистикой очищен!')

# Добавляем запросы в статистику
def write_file(op):
    file = open('history.txt', 'a', encoding= 'UTF-8') # 'a' - Добавить - Открывает файл для добавления, создает файл, если он не существует
    if op != '':
        file.write(op+'\n')
    file.close()

# Проверка на самый подходящий запрос функций(Почитайте про ratio и fuzzywuzzy)
def check(qwe):
    global command, i, write_file
    answer = ''
    for i in range(len(command)):
        case = fuzz.ratio(qwe,command[i])
        if (case > 50) & (case > i):
            answer = command[i]
            i = case
    if (answer != 'пока') & (answer != 'привет'):
        write_file(answer)
    return(str(answer))

# Работаем по поискам веб-сайтов, пускай пока будет Google
def web_browser():
    global adr
    webbrowser.open('https://www.google.com'.format(adr))

# Проверка на поиск в интернете
def check_browser_searching():
    global text, wifi_name, write_file
    global adr
    global web_browser
    if 'найди' in text:
        write_file('найди')
        adr = text.replace('найди','').strip()
        text = text.replace(adr,'').strip()
        check_browser_searching()
        text = ''
    elif 'найти' in text:
        write_file('найти')
        adr = text.replace('найти','').strip()
        text = text.replace(adr,'').strip()
        check_browser_searching()
        text = ''
    adr = ''

# Избавляемся от наших ключевых слов
def clear_keywords():
    global text, name_voice
    for i in name_voice:
        text = text.replace(i,'').strip()
        text = text.replace('  ',' ').strip()

# Приветствие 
def hello_margo():
    global engine
    hello = ['Привет, я ваш голосовой помощник. Иногда я торможу. Я еще в разработке. Ну что приступим?','Привет','Я рада вас снова слышать','Приветствую мой повелитель','Вечер в хату','Привет, чем могу быть полезна?','Да, я вас слушаю','Чем могу помочь?']
    z = random.choices(hello)
    engine.say(z)

# Выход из программы
def exit():
    global engine
    gb = ['хорошо, до новых встреч, пока','хорошо, до встечи','всего наилучшего','Была рада вам помочь','всегда к вашим услугам, Марго','до новых встреч мой господин']
    engine.say(random.choices(gb))
    engine.runAndWait();
    engine.stop();
    os.system('clear') # ИЛИ clear
    exit(0)

# Показ набор команд голосового ассистента
def list_cmd():
    my_commands = ['привет','выключи компьютер','включи youtube','открой браузер','какая сейчас погода','покажи файл статистик','переведи']
    for i in my_commands:
        print(i)
    time.sleep(2)

# Открываем браузер
def browser():
    webbrowser.open('https://www.google.com')

# Открываем Ютуб
def youtube():
    webbrowser.open('https://www.youtube.com')

# Оправляем ПК спать
def shutdown():
    global exit
    os.system('shutdown \s \f \t 15')
    exit()

# Погода
def weather():
    webbrowser.open('https://yandex.ru/pogoda')

# Открываем Vk
def vk():
    webbrowser.open('https://www.vk.com')

# Открываем Instagram
def instagram():
    webbrowser.open('https://www.instagram.com')

# Открываем StackOverflow
def stackover():
    webbrowser.open('https://ru.stackoverflow.com/')

# Перевод слов и текста
'''def translete():
    global text, ind
    ind = 0
    var = [' ка','переведи пожалуйста','перевод','переводить','переведи']
    for i in var:
        if (i in text)&(ind == 0):
            word = text
            word = word.replace('переведи','').strip()
            word = word.replace('перевести','').strip()
            word = word.replace('перевод','').strip()
            word = word.replace('переведи ка','').strip()
            word = word.replace('переведи пожалуйста','').strip()
            word = word.replace('слова','').strip()
            word = word.replace('слово','').strip()
            webbrowser.open('https://translate.yandex.ru/'.format(word))
            ind = 1
            text = ''
'''

cmds = {
    'Привет' : hello_margo,                  'здорова' : hello_margo,                 'выключи компьютер' : shutdown,
    'выключи комп' : shutdown,               'включи youtube' : youtube,              'открой браузер' : web_browser,
    'открой google' : web_browser,           'открой список команд' : list_cmd,       'какие команды ты знаешь' : list_cmd,
    'открой vk' : vk,                        'открой instagram' : instagram,          'открой stackoverflow' : stackover,
    'покажи погоду' : weather,               'какая сейчас погода' : weather,         'покажи погоду на завтра' : weather,
    'очисти файл' : clear_static,            'очисти файл статистики' : clear_static, 'покажи статистику' : clear_static,
    'покажи файл со статистикой' : history,  'покажи файл статистики' : history,      'покажи стату' : history,
    ''' 'переведи' : translete,                  'переведи пожалуйста' : translete,'''       'какие планы' : plane,
    'пока' : exit,                           'до встречи' : exit,                     'еще увидимся' : exit,
    'до скорого' : exit,                     'хорошего дня' : exit,                   'хорошего вечера' : exit,
    'добрый день' : hello_margo,             'добрый вечер' : hello_margo,            'привет марго' : hello_margo,
}


# распознавание речи
def talk():
    global text, clear_keywords
    text = ''
    with sr.Microphone() as source:
        print('Я вас слушаю: ')
        r.adjust_for_ambient_noise(source)
        audio = r.listen(source, phrase_time_limit=3)
        try:
            text = (r.recognize_google(audio, language="ru-RU")).lower()
        except(sr.UnknownValueError):
            pass
        except(TypeError):
            pass
        except (sr.UnknownValueError):
            speak("Я вас не поняла")
        os.system('clear')
        #lb['text'] = text # Это для окна приложения в будущем через tkinter
        clear_keywords()

# Поехали
def cmd_start():
    global cmds, engine, check, check_browser_searching, task_number, text #ld
    check_browser_searching()
    text = check(text)
    print(text)
    check_browser_searching()
    if (text in cmds):
        if (text != 'привет')&(text != 'пока')&(text != 'открой список команд'):
            case = ['Секундочку','Сейчас','Вот ваш результат','Выполняю',]
            engine.say(random.choices(case))
        cmds[text]()
    elif text == '':
        pass
    task_number += 1
    if (task_number % 10 == 0):
        engine.say('Что-нибудь еще?')
        engine.runAndWait()
        engine.stop()

# Основной бесконечный цикл
def main():
    global text, talk, cmd_start, i
    try:
        talk()
        if text != '':
            cmd_start()
            i = 0
    except(UnboundLocalError):
        pass
    except(TypeError):
        pass

while True:
    main()

Main_version_0.0/main.py
# BETA ВЕРСИЯ ГОЛОСОВОГО ПОМОЩНИКА МАРГО V 1.0

# импортирование библиотек
import speech_recognition as sr
import os
import sys
import webbrowser

def talk(words):
    print(words)
    os.system("say " + words)

talk("Привет, я ваш голосовой помощник. Иногда я торможу, потому что я еще в разработке. Ну что приступим?")

def command():
    r = sr.Recognizer()

    with sr.Microphone() as source:
        print("Говорите")
        r.pause_threshold = 1
        r.adjust_for_ambient_noise(source, duration=1)
        audio = r.listen(source)
    try:
        tz = r.recognize_google(audio, language="ru-RU").lower()
        print("Вы сказали " + tz)
    except sr.UnknownValueError:
        talk("Я вас не поняла")
        tz = command()

    return tz

def makeSomething(tz):
    if 'открой google' in tz:
        talk("Секундочку")
        url_0 = 'https://www.google.com'
        webbrowser.open(url_0)
    elif 'открой яндекс' in tz:
        talk("Секундочку")
        url_1 = 'https://yandex.ru/'
        webbrowser.open(url_1)
    elif 'как тебя зовут' in tz:
        talk("Меня зовут Марго")
    elif 'что ты умеешь делать' in tz:
        talk("Я могу назвать свое имя и открывать веб-сайты")
    elif 'марго' in tz:
        talk("Я вас слушаю, говорите")
    elif 'спасибо стоп' in tz:
        talk("Хорошо, до новых встреч, пока")
        sys.exit()

while True:
    makeSomething(command())


VoiceHelper-Home/main_v.0.1.py
# импортирование библиотек
import speech_recognition as sr
import os
import sys
import webbrowser
import random
import pyttsx3
from tkinter import *
import pandas as pd
from fuzzywuzzy import fuzz
import time
from datetime import datetime
from colorama import *

# Глобальные переменные
text = '' # Весь текст
r = sr.Recognizer() # Микрофон на устройсте, который слушает речь
engine = pyttsx3.init(); # Движок
adr = '' # Google браузер
i = 0 # для счетчика
task_number = 0


def speak(words):
    print(words)
    os.system("say " + words)


# Словарь приветствия Марго
name_voice = ['марго', 'маргарита', 'рита', 'ритка', 'маргунка', 'подруга' ,'не могла бы ты', 'пожалуйста', 'сделай услугу', 'покажи ка','покажи ']
# Словарь возможных команд (должно быть много пока без ИИ)
command = ['привет','здорова','выключи компьютер','выключи комп','включи youtube','открой браузер','открой google','открой список команд','какие команды ты знаешь','открой vk','открой instagram','открой stackoverflow','покажи погоду','какая сейчас погода','покажи погоду на завтра','очисти файл','очисти файл статистики','покажи статистику','покажи файл со статистикой','покажи файл статистики','покажи стату','какие планы','пока','до встречи','еще увидимся','до скорого','хорошего дня','хорошего вечера']

# Раздел функций с описанием команд Марго(Почитайте про Append и как работает библиотека pandas [53 строка pd.DataFrame] !!!!! )

# История запросов (review)
def history():
    z = {}
    arr = {}
    arr2 = {}
    arr3 = {}
    arr4 = {}
    file = open('history.txt', 'r', encoding = 'UTF-8')
    case = file.readlines()
    for i in range(len(case)):
        line = str(case[i].replace('\n','').strip())
        arr.append(line)
    file.close()
    for i in range(len(arr)):
        x = arr[i]
        if x in z:
            z[x] += 1
        if not(x in z):
            b = {x : 1}
            z.update(b)
        if not(x in arr2):
            arr2.append(x)
    for i in arr2:
        arr3.append(z[i])
    for i in range(1, len(arr3)+1):
        arr4.append(str(i)+') ')
    list = pd.DataFrame({
        'command' : arr2,
        'count' : arr3
    }, index = arr4)
    list.index.name = '№'
    print(list)

# Планы на проект Марго 1.0
def plane():
    global engine
    plane = '''Привет, я ваш голосовой помощник. Иногда я торможу, потому что я еще в разработке. В будующем я планирую быть конкурентом и возможно ассистентом вашего умного дома. Начиная от вашей технике, заканчивая вашим автомобилем и бытовой техникой. Пока еще у меня не было нормальных тестов, так что не ругайтесь, я все еще учусь.'''
    engine.say(plane)

# Очистка файла со статистикой
def clear_static():
    global engine
    file = open('history.txt', 'w', encoding= 'UTF-8') # "w" - Запись - Открывает файл для записи, создает файл, если он не существует
    file.close()
    engine.say('Ваш файл со статистикой очищен!')

# Добавляем запросы в статистику
def write_file(op):
    file = open('history.txt', 'a', encoding= 'UTF-8') # 'a' - Добавить - Открывает файл для добавления, создает файл, если он не существует
    if op != '':
        file.write(op+'\n')
    file.close()

# Проверка на самый подходящий запрос функций(Почитайте про ratio и fuzzywuzzy)
def check(qwe):
    global command, i, write_file
    answer = ''
    for i in range(len(command)):
        case = fuzz.ratio(qwe,command[i])
        if (case > 50) & (case > i):
            answer = command[i]
            i = case
    if (answer != 'пока') & (answer != 'привет'):
        write_file(answer)
    return(str(answer))

# Работаем по поискам веб-сайтов, пускай пока будет Google
def web_browser():
    global adr
    webbrowser.open('https://www.google.com'.format(adr))

# Проверка на поиск в интернете
def check_browser_searching():
    global text, wifi_name, write_file
    global adr
    global web_browser
    if 'найди' in text:
        write_file('найди')
        adr = text.replace('найди','').strip()
        text = text.replace(adr,'').strip()
        check_browser_searching()
        text = ''
    elif 'найти' in text:
        write_file('найти')
        adr = text.replace('найти','').strip()
        text = text.replace(adr,'').strip()
        check_browser_searching()
        text = ''
    adr = ''

# Избавляемся от наших ключевых слов
def clear_keywords():
    global text, name_voice
    for i in name_voice:
        text = text.replace(i,'').strip()
        text = text.replace('  ',' ').strip()

# Приветствие 
def hello_margo():
    global engine
    hello = ['Привет, я ваш голосовой помощник. Иногда я торможу. Я еще в разработке. Ну что приступим?','Привет','Я рада вас снова слышать','Приветствую мой повелитель','Вечер в хату','Привет, чем могу быть полезна?','Да, я вас слушаю','Чем могу помочь?']
    z = random.choices(hello)
    engine.say(z)

# Выход из программы
def exit():
    global engine
    gb = ['хорошо, до новых встреч, пока','хорошо, до встечи','всего наилучшего','Была рада вам помочь','всегда к вашим услугам, Марго','до новых встреч мой господин']
    engine.say(random.choices(gb))
    engine.runAndWait();
    engine.stop();
    os.system('clear') # ИЛИ clear
    exit(0)

# Показ набор команд голосового ассистента
def list_cmd():
    my_commands = ['привет','выключи компьютер','включи youtube','открой браузер','какая сейчас погода','покажи файл статистик','переведи']
    for i in my_commands:
        print(i)
    time.sleep(2)

# Открываем браузер
def browser():
    webbrowser.open('https://www.google.com')

# Открываем Ютуб
def youtube():
    webbrowser.open('https://www.youtube.com')

# Оправляем ПК спать
def shutdown():
    global exit
    os.system('shutdown \s \f \t 15')
    exit()

# Погода
def weather():
    webbrowser.open('https://yandex.ru/pogoda')

# Открываем Vk
def vk():
    webbrowser.open('https://www.vk.com')

# Открываем Instagram
def instagram():
    webbrowser.open('https://www.instagram.com')

# Открываем StackOverflow
def stackover():
    webbrowser.open('https://ru.stackoverflow.com/')

# Перевод слов и текста
def translete():
    global text, ind
    ind = 0
    var = [' ка','переведи пожалуйста','перевод','переводить','переведи']
    for i in var:
        if (i in text)&(ind == 0):
            word = text
            word = word.replace('переведи','').strip()
            word = word.replace('перевести','').strip()
            word = word.replace('перевод','').strip()
            word = word.replace('переведи ка','').strip()
            word = word.replace('переведи пожалуйста','').strip()
            word = word.replace('слова','').strip()
            word = word.replace('слово','').strip()
            webbrowser.open('https://translate.yandex.ru/'.format(word))
            ind = 1
            text = ''


cmds = {
    'привет' : hello_margo,                  'здорова' : hello_margo,                 'выключи компьютер' : shutdown,
    'выключи комп' : shutdown,               'включи youtube' : youtube,              'открой браузер' : web_browser,
    'открой google' : web_browser,           'открой список команд' : list_cmd,       'какие команды ты знаешь' : list_cmd,
    'открой vk' : vk,                        'открой instagram' : instagram,          'открой stackoverflow' : stackover,
    'покажи погоду' : weather,               'какая сейчас погода' : weather,         'покажи погоду на завтра' : weather,
    'очисти файл' : clear_static,            'очисти файл статистики' : clear_static, 'покажи статистику' : clear_static,
    'покажи файл со статистикой' : history,  'покажи файл статистики' : history,      'покажи стату' : history,
    ''' 'переведи' : translete,                  'переведи пожалуйста' : translete,'''       'какие планы' : plane,
    'пока' : exit,                           'до встречи' : exit,                     'еще увидимся' : exit,
    'до скорого' : exit,                     'хорошего дня' : exit,                   'хорошего вечера' : exit,
    'добрый день' : hello_margo,             'добрый вечер' : hello_margo,            'привет марго' : hello_margo,
}


# распознавание речи
def talk():
    global text, clear_keywords
    text = ''
    with sr.Microphone() as source:
        print('Я вас слушаю: ')
        r.adjust_for_ambient_noise(source)
        audio = r.listen(source, phrase_time_limit=3)
        try:
            text = (r.recognize_google(audio, language="ru-RU")).lower()
        except(sr.UnknownValueError):
            pass
        except(TypeError):
            pass
        except (sr.UnknownValueError):
            speak("Я вас не поняла")
        os.system('clear')
        #lb['text'] = text # Это для окна приложения в будущем через tkinter
        clear_keywords()

# Поехали
def cmd_start():
    global cmds, engine, check, check_browser_searching, task_number, text #ld
    check_browser_searching()
    text = check(text)
    print(text)
    check_browser_searching()
    if (text in cmds):
        if (text != 'привет')&(text != 'пока')&(text != 'открой список команд'):
            case = ['Секундочку','Сейчас','Вот ваш результат','Выполняю',]
            engine.say(random.choices(case))
        cmds[text]()
    elif text == '':
        pass
    task_number += 1
    if (task_number % 10 == 0):
        engine.say('Что-нибудь еще?')
        engine.runAndWait()
        engine.stop()

# Основной бесконечный цикл
def main():
    global text, talk, cmd_start, i
    try:
        talk()
        if text != '':
            cmd_start()
            i = 0
    except(UnboundLocalError):
        pass
    except(TypeError):
        pass

while True:
    main()